url3f8d6440dd_huiwq1990_CognativeModel_tgz-pJ8-clarion_samples_HelloWorldJ8/src/clarion/system/CLARION.java

```diff
@@ -0,0 +1,783 @@
+	 * feedback. 
+	 * <p>
+	 * If you are using the drive equation (or its equivalent) for calculating drive strengths within 
+	 * the MS, the sensory information should also include a dimension with the ID "STIMULUS" that 
+	 * contains the stimulus values for each of the drives. The stimulus for the drives is simply a 
+	 * translation from the sensory information (or context) into a single value to be used 
+	 * by the drive equation (etc.) for calculating the drive strength.
+	 * @param NewSensoryInformation The new sensory information to be perceived.
+	 * @param feedback The feedback received for performing the last action.
+	 * @throws MissingACSException If no instance of the ACS has been initialized.
+	 */
+	public void perceive (DimensionValueCollection NewSensoryInformation, double feedback) throws MissingACSException
+	{
+		Long ts = Birthdate + InternalClock;
+		if(acs == null)
+			throw new MissingACSException ("This instance of CLARION does not contain an ACS. At " +
+					"a MINIMUM, every instance of CLARION MUST contain an ACS.");
+		DimensionValueCollection PreInput = generatePreInput(NewSensoryInformation);
+		if(mcs != null)
+		{
+			Collection <AbstractMetaCognitiveModule> m = mcs.getModules(InterfaceMCSRunsAtPerception.class);
+			if(m != null)
+			{
+				for(AbstractMetaCognitiveModule a : m)
+				{
+					if(a.checkEligibility())
+						a.performMetaCognition(PreInput, ts);
+				}
+			}
+		}
+		if(gs != null)
+			CurrentGoal = gs.getCurrentGoal();
+		DimensionValueCollection NewInput = generateCurrentInput(NewSensoryInformation);
+		if(em != null)
+		{
+			em.add(NewInput, ts, EpisodicMemory.InfoStored.NEW_STATE);
+			em.addFeedback(feedback, ts);
+		}
+		if(mcs != null)
+		{
+			Collection <AbstractMetaCognitiveModule> m = mcs.getModules(InterfaceMCSRunsAfterEarlyPerceptionStage.class);
+			if(m != null)
+			{
+				for(AbstractMetaCognitiveModule a : m)
+				{
+					if(a.checkEligibility())
+						a.performMetaCognition(NewInput, ts);
+				}
+			}
+		}
+		acs.learn(NewInput, feedback, ts);
+		CurrentInput = NewInput;
+		if(mcs != null)
+		{
+			Collection <AbstractMetaCognitiveModule> m = mcs.getModules(InterfaceMCSRunsAfterLearning.class);
+			if(m != null)
+			{
+				for(AbstractMetaCognitiveModule a : m)
+				{
+					if(a.checkEligibility())
+						a.performMetaCognition(CurrentInput, ts);
+				}
+			}
+		}
+		++InternalClock;
+	}
+	
+	/**
+	 * Determines the action to perform given what was perceived. If this instance of CLARION
+	 * does not contain an ACS or no sensory information has been perceived (by calling the perceive
+	 * method) this method will throw an exception.
+	 * @return The performed action.
+	 * @throws MissingACSException If no instance of the ACS has been initialized.
+	 * @throws MissingSensoryInformationException If no sensory information has been perceived.
+	 */
+	public AbstractAction act ()  throws MissingACSException, MissingSensoryInformationException
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 20. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 284-295 into a separate method.